app/report_requests/process_notifications_report.py

```diff
@@ -0,0 +1,159 @@
+import csv
+from io import StringIO
+
+from flask import current_app
+from notifications_utils.s3 import (
+    S3_MULTIPART_UPLOAD_MIN_PART_SIZE,
+    s3_multipart_upload_abort,
+    s3_multipart_upload_complete,
+    s3_multipart_upload_create,
+    s3_multipart_upload_part,
+)
+
+from app.constants import NOTIFICATION_REPORT_REQUEST_MAPPING
+from app.dao.notifications_dao import get_notifications_for_service
+from app.dao.report_requests_dao import dao_get_report_request_by_id
+from app.dao.service_data_retention_dao import fetch_service_data_retention_by_notification_type
+
+
+def process_report_request(service_id, report_request_id):
+    report_request = dao_get_report_request_by_id(service_id, report_request_id)
+    notification_type = report_request.parameter["notification_type"]
+    notification_status = report_request.parameter["notification_status"]
+
+    page_size = current_app.config.get("REPORT_REQUEST_NOTIFICATIONS_CSV_BATCH_SIZE")
+    s3_bucket = current_app.config["S3_BUCKET_REPORT_REQUESTS_DOWNLOAD"]
+    filename = f"notifications_report/{report_request_id}.csv"
+    part_number = 1
+
+    init_upload_response = s3_multipart_upload_create(s3_bucket, filename)
+    upload_id = init_upload_response["UploadId"]
+    parts = []
+
+    csv_buffer = StringIO()
+    csv_writer = csv.writer(csv_buffer)
+
+    headers = [
+        "Recipient",
+        "Reference",
+        "Template",
+        "Type",
+        "Sent by",
+        "Sent by email",
+        "Job",
+        "Status",
+        "Time",
+        "API key name",
+    ]
+    csv_writer.writerow(headers)
+
+    try:
+        service_retention = fetch_service_data_retention_by_notification_type(service_id, notification_type)
+        limit_days = service_retention.days_of_retention if service_retention else 7
+        page = 1
+
+        while True:
+            serialized_notifications = get_notifications_by_batch(
+                service_id=service_id,
+                template_type=notification_type,
+                notification_status=notification_status,
+                page=page,
+                page_size=page_size,
+                limit_days=limit_days,
+            )
+
+            csv_data = convert_notifications_to_csv(serialized_notifications)
+            csv_writer.writerows(csv_data)
```

Comment:
we need to re-use the buffer because we're using the same `csvwriter`. but since we're not using a more advanced writer (ie: we don't need to keep track of headers), we could just create a new csv writer each time, and with that create a new buffer each time so it's simpler to manage and we would never need to `seek` at all